Coding runtime resources

aperezdc opened the pull request:
This is a first PR which adds the basics to decide in which mode to run (normal vs. coding), by checking the file name of the executable. Renaming the installer as `endless-coding.exe` will use coding mode. The first commit just adds an `CEndlessUsbToolDgl::IsCoding()` static method, and the second one chooses the URLs and file names for the JSON live/installer images using the new function.

(I will be submitting a series of small PRs starting from this one. The idea is to ease code review, and to ensure that at any intermediate point the code is still building fine.) 

---

https://phabricator.endlessm.com/T14033

---

Checklist to help myself keep tabs on what is done:

- [x] Detection based on executable name:
  - [x] Detection of `-code-` string in name (enables coding mode).
  - [x] Decoupling of uninstaller detection (allows detecting of coding mode for the uninstaller).
- [x] Analytics: Sends a `cd2` custom dimension with values `eos` or `code`.
- [x] Installer UI:
  - [x] [C++ format strings] All the occurrences of `Endless OS` have an `Endless Code` alternative.
  - [x] [HTML] All the occurrences of `Endless OS` in the HTML have an `Endless Code` in the HTML have alternatives, saving for `ThankYouDualBootReminder` to avoid needing a different `/etc/os-release`, etc.
  - [x] [HTML] It is possible to show/hide elements using CSS classes `.ModeCoding` and `.ModeNormal`.
  - [ ] [HTML] Alternate bitmap(s) for coding mode → Doable with the CSS classes, Code artwork for the installer is not available at the moment.
- [ ] Image handling:
  - [x] Fetch alternate JSON containing list of images.
  - [ ] Hiding the “Light” image selection for coding mode.
  - [x] Skipping the image selection page when there is only one image.

aperezdc committed:
Coding: Check whether we are called coding-installer.exe

aperezdc committed:
Coding: Move switcheroo for JSON files/URLs to helper functions

Instead of making each location where expansions of the JSON_LIVE_FILE and
JSON_INSTALLER_FILE also check whether we are in "Coding Mode", add helper
functions which always return the relevant value, handling both the cases
of ENABLE_JSON_COMPRESSION being enabled, and being in "Coding Mode". The
functions are:

 - JsonLiveFile()
 - JsonLiveFileURL()
 - JsonInstallerFile()
 - JsonInstallerFileURL()

All the expansions of the JSON_LIVE_FILE, JSON_INSTALLER_FILE, JSON_URL, and
JSON_PACKED macros have been replaced to use the helper functions. Code added
in the future shall use the helper functions.

aperezdc committed:
Coding: Use CSS to show/hide elements depending on the mode

This calls the setCodingMode() JavaScript function to allow showing/hiding
elements from the HTML UI depending on the operating mode. The function
adds either the "ModeNormal" or "ModeCoding" classes to the <body> element,
and CSS rules which hide the elements nested inside with a non-matching
class. Elements to be shown in both modes do not need any CSS class.

aperezdc committed:
Coding: Add l18n resources for Coding mode strings

aperezdc committed:
Coding: Switcheroo for #SelectStorageSubtitle element

aperezdc committed:
Coding: Switcheroo for #StorageSpaceDescription element

wjt commented on src/endless/EndlessUsbToolDlg.cpp:
```diff
@@ -619,6 +620,7 @@ void CEndlessUsbToolDlg::OnDocumentComplete(LPDISPATCH pDisp, LPCTSTR szUrl)
         CallJavascript(_T(JS_ENABLE_BUTTON), CComVariant(HTML_BUTTON_ID(_T(ELEMENT_REFORMATTER_USB_BUTTON))), CComVariant(FALSE));
     }
 
+	CallJavascript(_T(JS_SET_CODING_MODE), CComVariant(IsCoding()), CComVariant(NULL));
```
The `CComVariant(NULL)` here is not needed, the parameter has a default value.

wjt commented on src/endless/res/EndlessUsbTool.htm:
```diff
@@ -307,9 +312,12 @@
-                            <div class="ReminderTitle" id="ThankYouLiveReminder">To try Endless OS now, restart your computer.</div>
-                            <div class="ReminderTitle" id="ThankYouReflasherReminder">Now you're ready to install Endless OS on more computers.</div>
+                            <div class="ReminderTitle ModeNormal" id="ThankYouDualBootReminder">Restart your computer and select "Endless OS" from the menu.</div>
+                            <div class="ReminderTitle ModeCoding" id="ThankYouDualBootReminderCoding">Restart your computer and select "Endless Code" from the menu.</div>
```
Have you arranged for "Endless Code" to appear in the GRUB menu? This string comes ultimately from the `/boot/loader/...` entry (read by our GRUB's `bls_import`), which I guess must be created in the OSTree builder, though I can't find it. @dbnicholson do you know where that file is generated?

aperezdc replied:
I couldn't find this either, so for now I only have this bit here to change the string in the installer.

aperezdc replied:
FWIW, I have found one occurrence in [eos-image-builder:data/grub.cfg](https://github.com/endlessm/eos-image-builder/blob/master/data/grub.cfg#L108-L110), but it is a menu entry just showing a warning message about Endless OS not being available while Windows is hibernated. :rofl:

wjt replied:
It comes from `/etc/os-release`: https://github.com/endlessm/ostree/blob/923e6dd80fcc15084295932502b2aeab4a96301f/src/libostree/ostree-sysroot-deploy.c#L1410

This is created in `base-files`, I think: `https://github.com/endlessm/base-files/blob/84f19f81c562ef58f7d956f8a393984422d818ab/src/os-release.in`

It's already patched by the ostree builder to include the build ID so you could conceivably arrange for it to be patched with the product name there: https://github.com/endlessm/eos-ostree-builder/blob/master/hooks/os/60-image_version

Maybe ask dsd or dbn?

wjt commented on src/endless/res/endless.loc:
```diff
@@ -265,6 +265,7 @@ t MSG_339 "%s (Recommended)"
+t MSG_345 "Select how much of your %s drive to dedicate to Endless Code."
 
 t MSG_350 "Unfortunately, there is not enough storage space."
 t MSG_351 "To continue installing Endless OS, please free up at least %s by deleting videos, photos, or other large files on your %s drive."
```
How much do you care about “Endless OS” in error strings? Maybe it's not so bad, at least for now.

I guess in theory we could replace them *all* with `%s` and sub the product name in at runtime, but this would probably be very confusing to translators.

wjt replied:
One option would be to take a leaf out of [Rufus' book](https://github.com/endlessm/rufus/blob/master/src/localization.c#L409) and use one of the high bits to mark strings as Code-flavoured; then have some `uprintf` wrapper which, if in Code mode, checks whether a high-bit variant of a string exists and uses it instead. I don't think it's worth the complexity for just a few strings, but maybe if the number grows...

aperezdc replied:
For now it's not a lot of strings, and if we leave the error strings out, it would be even less... I think we can live with a few “duplicated” strings for now.

wjt commented on src/endless/EndlessUsbToolDlg.cpp:
```diff
@@ -297,19 +297,21 @@ enum endless_action_type {
+#define JSON_LIVE_FILE             "releases-eos-3.json"
+#define JSON_CODING_LIVE_FILE      "releases-coding-3.json"
+#define JSON_INSTALLER_FILE        "releases-eosinstaller-3.json"
+#define JSON_CODING_INSTALLER_FILE "releases-codinginstaller-3.json"
```
I don't think you actually want to build `codinginstaller` images; we should always use the `eosinstaller` images.

Some background: once upon a time, you had to choose. A live USB was just a normal image written to a USB, and you couldn't install from it. An installer USB was an `eosinstaller` image with an extra exFAT partition containing image file(s) to install; but you couldn't use this as a live USB to try the image you're about to install.

Now, there's a "combined" USB which can do both. This is achieved (essentially) by loopback-mounting the uncompressed, pristine image from an exFAT partition, and including the `eos-installer` binary (note the hyphen!) inside all images. In this model, the separate `eosinstaller` image is not used.

Unless you [hold the `Control` key while clicking the “advanced options” link, or are running Windows XP](https://github.com/endlessm/rufus/blob/master/src/localization.c#L409), this app only gives you a way to create new-style combined USBs.

Given that `eosinstaller` images are only used in a “cheat mode”, I don't think you want to introduce a `codinginstaller` product.

aperezdc replied:
This is a very insightful clarification, thanks! I'll remove the `codinginstaller`-fu.

wjt commented on src/endless/EndlessUsbToolDlg.cpp:
```diff
@@ -321,6 +324,7 @@ const wchar_t* mainWindowTitle = L"Endless Installer";
 
 #define ALL_FILES					L"*.*"
 
+#define CODING_INSTALLER_NAME    L"coding-installer.exe"
```
Oh yeah, one more thing: if the brand is "Endless Code", shouldn't this be `code-installer.exe` or `endless-code-installer.exe`?

Also, at present you'll have "Endless OS" strings in the uninstaller.

aperezdc replied:
Yep, I have to change this. Actually, I was thinking that it would be good to check for the `code-installer` prefix, and `.exe` suffix. The idea is that sometimes users can be a bit goofy, and if they download the same file twice by accident, they may end up with several files with names like:

* `code-installer.exe`
* `code-installer(1).exe`
* `code-installer.1.exe`

...and so on, depending on which browser or download manager they use to fetch the file. Any of the names above should trigger coding mode, IMO.

aperezdc replied:
As for the uninstaller, I remember talking about that being fine for a first pass, and that if needed we would address that with a separate PR.

wjt replied:
You could check for the presence/absence of `-code-` in the filename, perhaps. And maybe adapt the uninstall path to just check for the presence of `-uninstaller`. This would allow all four states of {installer, uninstaller} × {normal, code}

aperezdc commented:
On Wed, 11 Jan 2017 01:29:13 -0800, Will Thompson <notifications@github.com> wrote:
> wjt commented on this pull request.
> 
> 
> 
> > @@ -321,6 +324,7 @@ const wchar_t* mainWindowTitle = L"Endless Installer";
>  
>  #define ALL_FILES					L"*.*"
>  
> +#define CODING_INSTALLER_NAME    L"coding-installer.exe"
> 
> You could check for the presence/absence of `-code-` in the filename, perhaps. And maybe adapt the uninstall path to just check for the presence of `-uninstaller`. This would allow all four states of {installer, uninstaller} × {normal, code}

I like this idea. And I already have the check for `-code-` in place. Now on
to adding the check for `-uninstaller.exe` (in this case I think we can have a
check for the exact suffix, instead of searching for a contained substring,
because we know under which file name the installer is copied so it will act
as an uninstaller).

aperezdc committed:
Check only for -uninstaller.exe suffix in IsUninstaller()

This makes it possible to use a different file name for the uninstaller in
"coding mode", and have both IsCoding() and IsUninstaller() return "true".

aperezdc committed:
Pass mode (eos/code) as an Analytics custom dimension

dbnicholson commented:
If you want something different in /etc/os-release then you need a separate ostree for coding, which is something @cosimoc did not want.

wjt commented:
That's fine, of course. The consequence is just that the GRUB menu will
show “Endless OS”, and so we should keep the same
`ThankYouDualBootReminder` text ("Restart your computer and select
\"Endless OS\" from the menu.") rather than having a Code version.

aperezdc commented:
@dbnicholson @wjt: I'll be reverting the bit for `ThankYouDualBootReminder` then. Thanks for commenting on that!

wjt commented on src/endless/EndlessUsbToolDlg.cpp:
```diff
@@ -6084,11 +6082,79 @@ void CEndlessUsbToolDlg::DelayDeleteFolder(const CString &folder)
 	}
 }
 
+static bool _lowerCaseExeNameFindSubstring(CStringW exePath, const wchar_t* substring, bool* isSuffix)
```
This is declared to return `bool` but in fact returns an `int` and its two callers assume its return value is an `int`. MSVC emits the following warnings:

```
1>c:\users\wjt\source\repos\rufus\src\endless\endlessusbtooldlg.cpp(6094): warning C4800: 'int': forcing value to bool 'true' or 'false' (performance warning)
1>c:\users\wjt\source\repos\rufus\src\endless\endlessusbtooldlg.cpp(6113): warning C4804: '>': unsafe use of type 'bool' in operation
```

aperezdc replied:
Good catch :+1: , latest pushed version has this fixed.

wjt commented on src/endless/res/endless.loc:
```diff
@@ -101,23 +101,31 @@ g IDD_ENDLESSUSBTOOL_DIALOG
 t SelectFilePreviousButton "Previous"
 t SelectFileNextButton "Next"
 t SelectFilePageTitle "Select Your Endless OS Version"
+t SelectFilePageCoding "Select Your Endless Code Version"
```
```11:07:03 - embedded.loc(112): unknown control 'SelectFilePageCoding'```

aperezdc replied:
Fixed as well in the latest push.

wjt commented on src/endless/Analytics.cpp:
```diff
@@ -331,6 +332,7 @@ void Analytics::urlEncode(const CString &in, CString &out)
-		m_trackingId, m_clientId, _T(RELEASE_VER_STR), m_language, m_windowsVersion);
+	id.Format(_T("v=1&tid=%s&cid=%s&an=Endless%%20Installer&av=%s&ul=%s&cd1=%s&cd2=%s"),
+		m_trackingId, m_clientId, _T(RELEASE_VER_STR), m_language, m_windowsVersion,
+		CEndlessUsbToolDlg::IsCoding() ? "code" : "eos");
```
These need to be wide strings, and you're missing the trailing `&`:

```
11:07:03 - Analytics req: v=1&tid=UA-61784217-3&cid=ffa20047-8a44-4010-820b-223726d5dd26&an=Endless%20Installer&av=3.0.11.1-23-g3f8ba4a&ul=en-US&cd1=Windows%2010%2064-bit%20(Build%2014393)&cd2=??t=screenview&cd=DualBootInstallPage&sc=start
11:07:03 - Analytics: response code 200
11:07:03 - Analytics: submitted an invalid hit: {
  "hitParsingResult": [ {
    "valid": false,
    "parserMessage": [ {
      "messageType": "ERROR",
      "description": "A value is required for parameter 'dp'. Please see http://goo.gl/a8d4RP#dp for details.",
      "messageCode": "VALUE_REQUIRED",
      "parameter": "dp"
    } ],
    "hit": "/debug/collect?v=1\u0026tid=UA-61784217-3\u0026cid=ffa20047-8a44-4010-820b-223726d5dd26\u0026an=Endless Installer\u0026av=3.0.11.1-23-g3f8ba4a\u0026ul=en-US\u0026cd1=Windows 10 64-bit (Build 14393)\u0026cd2=??t=screenview\u0026cd=DualBootInstallPage\u0026sc=start"
  } ],
  "parserMessage": [ {
    "messageType": "INFO",
    "description": "Found 1 hit in the request."
  } ]
}
```

Notice in the URL: `&cd2=??t=screenview&`

aperezdc replied:
Right, I missed the fact that more parameters were being added. Also fixed in last push.

aperezdc committed:
Use always ThankYouDualBootReminder to avoid needing to change Grub labels

aperezdc commented:
I have added a checklist in the PR description to have a clear outline of what the PR is including. I am working on getting commits pushed for the remaining two items I can handle myself ­— unfortunately at the moment we don't have bitmap to show during installation for Code, so I'll have to leave that (it should be fairly easy to add later by reusing the CSS classes to hide the normal image and show the alternate one, though).

aperezdc committed:
Handle coding mode in C++ format strings

This reuses existing (but unused) localization string identifiers. Fortunately
there are not that many strings, so in the end it is not needed to change the
value of MSG_MAX.

aperezdc committed:
Comment MSG_303 as it's not used anywhere

aperezdc committed:
Add presumably missing "break;" in case block

aperezdc commented:
@wjt: Could you please take a look? I think it would be good to merge this PR in its current form because it's starting to be quite big, and the remaining two bits (skipping with only one image; and showing a different bitmap while installing) can be done in separate PRs.

wjt commented:
Rebased, will test and push any fixes I think this needs…

wjt committed:
Fetch coding image index from internal host

This also tidies up the building of index URLs a bit, reducing the number
of `#ifdef ENABLE_JSON_COMPRESSION` blocks.

wjt committed:
Skip SelectFilePage if exactly one image is available

In Coding mode, there is only one image (base). If we have no local images
to display, just skip the image selection page entirely.

If there's 0 or >1 remote image, or at least one local image, show the
page. If the manifest has not yet been fetched, we'll show the selection
page, and fill in the Basic half of the pane once it's fetched. This is
not ideal but will do for now.

wjt committed:
loc2pot: skip Endless Code strings

When we've finalized the design for Code Mode, we'll likely want to
change some strings, so let's save our translators some work.

wjt commented:
Branch updated. I dropped the patches that hid the "light" column in the image selection screen. Because the only `coding` image is `base`, there are no non-"light" images, so hiding that column made it impossible to select an image.

Instead, I chose to skip that page entirely if all the following are true:

* The image index has been retrieved
* There is exactly one image in it
* There are no local images

If the user is really fast and clicks Install before the index is fetched, then okay, they get the selection screen with only one functional column. This is good enough for now, I think, because it gives us a usable installer.

I also decided to filter Code-specific strings out of `endless.pot`. Until these get some design review I don't want to waste our translators' time!

ltilve commented:
I have been taking a look, and all the changes seem reasonable to me. I added @aperezdc to the review loop, to get also his approval.

aperezdc commented on src/endless/EndlessUsbToolDlg.cpp:
```diff
@@ -2066,11 +2066,10 @@ HRESULT CEndlessUsbToolDlg::OnLiveUsbClicked(IHTMLElement* pElement)
     FUNCTION_ENTER;
 
-	SetSelectedInstallMethod(InstallMethod_t::LiveUsb);
+    SetSelectedInstallMethod(InstallMethod_t::LiveUsb);
```
Maybe it would be possible to avoid this line appearing in the diff by avoiding the spacing changes. Or is this s CR vs. CRLF thing? At any rate, do not consider this thing a blocker — just a “nice to have” thing.

aperezdc replied:
There's a few more occurrences in the file. As said: if it's not annoying to do, it would be nice to get the diff to have only the actual changes (super low priority!)

wjt replied:
This is deliberate-ish. When I wrote the [.editorconfig](https://github.com/endlessm/rufus/blob/master/.editorconfig#L7-L10) file, I checked and this file had pretty much equal spaces and tabs. I try to be consistent with what's already there but err on the side of spaces. In these cases, these 3-4 line functions already had inconsistent whitespace (the `FUNCTION_ENTER`s are space-indented, the other stuff tab-indented) so I decided to fix them while I was there, but this is lazy of me… I'll rework the patch.

wjt replied:
Okay, updated: the whitespace changes are a separate commit.

aperezdc replied:
Awesome, thanks :tada:

aperezdc reviewed (commented):
The latest additions by @wjt look jut fine to me. I have commented about a small nit, but absolutely not a blocker.

wjt committed:
Fix indentation in a few callbacks
